File: src/services/instructions/loadstore/LD_B_REG_MXX.java
package emulator.source.loadstore

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class LD_B_REG_MXX extends Instruction {
	public LD_B_REG_MXX(memory, address, source, 
 destination) {
		super(memory, address, source, destination)
		super.setArgument32()
		super.setAssembler("ld.b " + this.sdest + ", [0x%08x]")
	}

	
	exec ({ context, memory }) {
		int fixedAddr = fix(this.argument)
		if ((fixedAddr & 1) == 0)
			context.getReg(this.destination)  = (short)(context.memory[fixedAddr / 2] >> 8) & 0xFF
		else
			context.getReg(this.destination)  = (short)(context.memory[fixedAddr / 2] & 255) & 0xFF
		context.pc  += 6
	}
}

File: src/services/instructions/loadstore/LdBRegMXX.js

 import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination) => `ld.b ${sdestination}, [0x%08x]`

export default class LdBRegMXX extends Instruction {
	constructor (memory, address, source, destination, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		super.setArgument32(memory)
		super.setAssembler(ASSEMBLER_INSTRUCTION_EXPRESSION(this.sdestination))
	}

	exec ({ context, memory }) {
		const fixedAddr = Instruction.fix(this.argument)
		if ((fixedAddr & 1) == 0) {
			// TODO -->> (short)(context.memory[fixedAddr / 2] >> 8) & 0xFF
			context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  = (memory[fixedAddr / 2] >> 8) & 0xFF
		} else {
			// TODO -->> (short)(context.memory[fixedAddr / 2] & 255) & 0xFF
			context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  = (memory[fixedAddr / 2] & 255) & 0xFF
		}

		context.pc  += 6
	}
}
